cmd/kube-scheduler/app/options/configfile.go

```diff
@@ -37,39 +40,61 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
 	return loadConfig(data)
 }
 
-func loadConfig(data []byte) (*kubeschedulerconfig.KubeSchedulerConfiguration, error) {
+func loadConfig(data []byte) (*config.KubeSchedulerConfiguration, error) {
 	// The UniversalDecoder runs defaulting and returns the internal type by default.
-	obj, gvk, err := kubeschedulerscheme.Codecs.UniversalDecoder().Decode(data, nil, nil)
+	obj, gvk, err := scheme.Codecs.UniversalDecoder().Decode(data, nil, nil)
 	if err != nil {
 		return nil, err
 	}
-	if cfgObj, ok := obj.(*kubeschedulerconfig.KubeSchedulerConfiguration); ok {
+	if cfgObj, ok := obj.(*config.KubeSchedulerConfiguration); ok {
+		// We don't set this field in pkg/scheduler/apis/config/{version}/conversion.go
+		// because the field will be cleared later by API machinery during
+		// conversion. To work around that, we set the API version field of the
+		// internal object here to the version we read from file. We need this for
+		// later when we need to make decisions based on the external version.
+		cfgObj.TypeMeta.APIVersion = gvk.GroupVersion().String()
 		return cfgObj, nil
 	}
 	return nil, fmt.Errorf("couldn't decode as KubeSchedulerConfiguration, got %s: ", gvk)
 }
 
+func encodeConfig(cfg *config.KubeSchedulerConfiguration) (*bytes.Buffer, error) {
+	buf := new(bytes.Buffer)
+	const mediaType = runtime.ContentTypeYAML
+	info, ok := runtime.SerializerInfoForMediaType(scheme.Codecs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return buf, fmt.Errorf("unable to locate encoder -- %q is not a supported media type", mediaType)
+	}
+
+	var encoder runtime.Encoder
+	switch cfg.TypeMeta.APIVersion {
```

Comment:
this is switching on encoding the internal config based on the external version it has in its apiVersion field? shouldn't we pick a default version to encode to (latest seems reasonable) like we did previously?

Reply:
since the internal object could've been converted from any version, there could be some information loss if we convert for example v1beta1 -> internal -> v1beta2.

to be quite honest, and I discussed this with @alculquicondor, I don't understand the purpose of this feature. who is using this and for what?

Reply:
> I don't understand the purpose of this feature. who is using this and for what?

isn't this the path taken when generating a config with `--write-config-to`?

Reply:
Yes.

Reply:
right, but I don't fully understand what that feature is used for. perhaps to generate a default config file that you can modify afterwards?

Reply:
> perhaps to generate a default config file that you can modify afterwards?

it's not just a default config file, it writes the effective configuration resulting from the flags you set, so someone transitioning from flags to a config file can add `--write-config-to=foo.yaml` to their existing flag-based invocation and get the equivalent config file.

Reply:
using latest version sgtm

Reply:
Done.

Reply:
actually, I'm having second thoughts about this: we necessarily need to switch when someone loads a v1beta1 config file that uses a deprecated plugin because the v1beta2 encoder cannot encode plugin args for deprecated plugins. am I missing something?

Reply:
I don't object to switching on APIVersion, my comment was limited to the behavior when APIVersion was empty (pick a default, instead of erroring)

Reply:
ah I completely misread your comment. done.